src/express.re

```diff
@@ -1,43 +1,149 @@
+
+  external params : t => params = "" [@@bs.get];
+  /** [params request] return the JSON object filled with the 
+      request parameters */
+
+  external asJSONObject : t => Js_dict.t Js_json.t = "%identity";
+  /** [asJSONObject request] casts a [request] to a JSON object. It is 
+      common in Express application to use the Request object as a 
+      placeholder to maintain state through the various middleware which 
+      are executed. */ 
+
+  /** more property method coming later */
 };
 
 module Response = {
   type t;
-  external sendFile : t => string => 'a => unit = "" [@@bs.send];
-  external sendString : t => string => unit = "send" [@@bs.send];
-  external sendJson : t => Js.Json.t => unit = "send" [@@bs.send];
-  external sendBuffer : t => Buffer.t => unit = "send" [@@bs.send];
-  external sendArray : t => array 'a => unit = "send" [@@bs.send];
-  external json : t => Js.Json.t => unit = "" [@@bs.send] [@@ocaml.deprecated "Use sendJson instead`"];
+  external sendFile : t => string => 'a => done_ = "" [@@bs.send];
+  external sendString : t => string => done_ = "send" [@@bs.send];
+  external sendJson : t => Js.Json.t => done_ = "send" [@@bs.send];
+  external sendBuffer : t => Buffer.t => done_ = "send" [@@bs.send];
+  external sendArray : t => array 'a => done_ = "send" [@@bs.send];
+  external json : t => Js.Json.t => done_ = "" 
+    [@@bs.send] [@@ocaml.deprecated "Use sendJson instead`"];
 };
 
-module Express = {
+module Next : {
+  type content;
+
+  type t = (Js.undefined content => done_) [@bs];
+
+  let router : Js.undefined content;  
+  /** value to use as [next] callback argument to skip middleware 
+      processing. */ 
+
+  let error : Error.t => Js.undefined content;
+  /** [error e] returns the argument for [next] callback to be propagate 
+      error [e] through the chain of middleware. */
+
+} = {
+
+  type content;
+
+  type t = (Js.undefined content => done_) [@bs];
+
+  external castToContent : 'a => content = "%identity";
+
+  let router = 
+    Js.Undefined.return (castToContent "router");
+
+  let error (e:Error.t) =>
+    Js.Undefined.return (castToContent e);
+
+};
+
+module Middleware = {
+  type next = Next.t;
+  type t; 
+      /* Middleware abstract type which unified the various way an
+         Express middleware can be constructed:
+         {ul
+         {- from a {b function} using the [ofFunction]}
+         {- from a Router class}
+         {- from an App class}
+         {- from the third party middleware modules}
+         }
+         For each of the class which implements the middleware interface
+         is JavaScript, one can use the "%identity" function to cast
+         it to this type [t] */
+  type f = Request.t => Response.t => next => done_;
+  
+  external fromF: f => t = "%identity"
+    [@@ocaml.doc " Create a Middleware from a function " ]; 
+  
+  type errorF = Error.t => Request.t => Response.t => next => done_;
+  
+  external fromErrorF: errorF => t = "%identity"
```

Comment:
I think `from` and `fromError` are descriptive enough. It's implicit from the context of the `Middleware` module that these are functions.